airflow/providers/edge/executors/edge_executor.py

```diff
@@ -0,0 +1,175 @@
+    ) -> None:
+        """Execute asynchronously."""
+        self.validate_airflow_tasks_run_command(command)
+        session.add(
+            EdgeJobModel(
+                dag_id=key.dag_id,
+                task_id=key.task_id,
+                run_id=key.run_id,
+                map_index=key.map_index,
+                try_number=key.try_number,
+                state=TaskInstanceState.QUEUED,
+                queue=queue or DEFAULT_QUEUE,
+                command=str(command),
+            )
+        )
+
+    @provide_session
+    def sync(self, session: Session = NEW_SESSION) -> None:
+        """Sync will get called periodically by the heartbeat method."""
+        jobs: list[EdgeJobModel] = session.query(EdgeJobModel).all()
+        for job in jobs:
+            if job.key in self.running:
+                if job.state == TaskInstanceState.RUNNING:
+                    if (
+                        job.key not in self.last_reported_state
+                        or self.last_reported_state[job.key] != job.state
+                    ):
+                        self.running_state(job.key)
+                    self.last_reported_state[job.key] = job.state
+                elif job.state == TaskInstanceState.SUCCESS:
+                    if job.key in self.last_reported_state:
+                        del self.last_reported_state[job.key]
+                    self.success(job.key)
+                elif job.state == TaskInstanceState.FAILED:
+                    if job.key in self.last_reported_state:
+                        del self.last_reported_state[job.key]
+                    self.fail(job.key)
+                else:
+                    self.last_reported_state[job.key] = job.state
+            job_success_purge = conf.getint("edge", "job_success_purge")
+            job_fail_purge = conf.getint("edge", "job_fail_purge")
+            if (
+                job.state == TaskInstanceState.SUCCESS
+                and job.last_update_t < (datetime.now() - timedelta(minutes=job_success_purge)).timestamp()
+            ) or (
+                job.state == TaskInstanceState.FAILED
+                and job.last_update_t < (datetime.now() - timedelta(minutes=job_fail_purge)).timestamp()
+            ):
+                if job.key in self.last_reported_state:
+                    del self.last_reported_state[job.key]
+                session.delete(job)
+                session.execute(
+                    delete(EdgeLogsModel).where(
+                        EdgeLogsModel.dag_id == job.dag_id,
+                        EdgeLogsModel.run_id == job.run_id,
+                        EdgeLogsModel.task_id == job.task_id,
+                        EdgeLogsModel.map_index == job.map_index,
+                        EdgeLogsModel.try_number == job.try_number,
+                    )
+                )
+        session.commit()
+
+    def end(self) -> None:
+        """End the executor."""
+        self.log.info("Shutting down EdgeExecutor")
```

Comment:
Shouldn't this do something to block and flush running tasks?

Reply:
I was also thinking about this. From the position of (potentially) multiple scheduler instances and the structure that open jobs are in the DB table... actually all schedulers contribute to the queue of jobs. No scheduler "owns" a job but just distributes them.

I understand that a graceful stop is needed for a Local Executor. But in the case of Edge the jobs are running remote. Stopping the scheduler will just end adding new jobs, existing (queued) jobs can continue also after stopping the scheduler. Is it a required interface to wait for all tasks? What if another scheduler is adding more jobs... then the scheduler going to stop would potentially never terminate. Or we need to tag which were the tasks created by this scheduler instance to know when which instance is eligible to "end".

For sake of simplicity I thought we can just stop.

If you think I need to drain all jobs started, then I would need to add a draining of (remote running) jobs. In my tests stopping the scheduler did not harm the completion (async). Let me know. (I am not an expert on the scheduler interface...
(Following the docs in https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/executor/index.html the implementation is optional though)

Reply:
> Or we need to tag which were the tasks created by this scheduler instance to know when which instance is eligible to "end".

Tasks are already owned by a specific scheduler, you will see dag runs and task instances all have the job_id of the scheduler job runner which started them. This is how failing stuck tasks and task adoption all works. If the job runner that ran the tasks dies/is killed (i.e. it doesn't heartbeat for some period of time) the TIs will be adopted by another scheduler job runner (if your executor has task adoption implemented, which this one doesn't) or be reset if they are not adopted.

So not mandatory but you will get funny stuff happening if you don't drain your tasks or at least support adoption. This executor is doing neither. But things will also work if your tasks are short enough to not hit that window.

Reply:
In my "Integration Test" DAG (that is alread ymerged from previous PRs and sitting in `airflow/providers/edge/example_dags/integration_test.py`) I added a long running task (15min) to ensure a heartbeat is tested. That all is running fine. Also thanks to `tenancity` added to communication layer a disconnect of ~5min is survived and task still is in success.

But your comment reminds me that I wanted to make two stability tests which I forgot:
1. Stop and restart the scheduler after some time and check that all tasks will run to finish
2. kill -9 the worker and task run process and ensure that after missing heartbeat the task is marked with error and leftovers are cleaned in the edge_job table.

I'll double check this... tomorrow.

Reply:
Hi @o-nikolas ... took a moment but now I was able to make the stability checks, this is the outcome:

Tests were based in the "Integration Test" DAG which contains all tested functions and operators which is added to provider package - will be base of system test in future if I get to this point. As you asked for heartbeat - there is one "long running" tasks that sleeps 1 minute and sends some logs for 15 minutes runtime - so a task that is hitting the heartbeat timeout. Added this before to ensure heartbeating is working and long running tasks can be made on the edge.

- [x] Positive case - let it run - SUCCESS - long running task reported OK
- [x] Kill -9 the "task run" process - FAILED as expected - task was reported as failed (immediately). Logs are expected like:
      `INFO - Task exited with return code -9For more information, see https://airflow.apache.org/docs/apache-airflow/stable/troubleshooting.html#LocalTaskJob-killed`
- [x] Kill -9 edge worker, leave forked long task process running - SUCCESS (because task completes also w/o worker and updates the state itself via task runner API)
      Side effect is that logs are not uploaded as this is done by the worker which is dead.
      After the restart of Edge Worker, the DAG picked the last task and ended with success.
      Future improvement will be to upload leftover logs after restart of edge worker
- [x] Kill -9 edge worker and task runner (like if a container or server dies - FAILED as expected
      After 300(+) seconds - the task was detected as zombie and was set to failed by the scheduler.
      No need to have something special implemented on the EdgeExecutor or Edge Worker (which was dead anyway)
- [x] Cut network connectivity (Actually stop webserver/API) - FAILED - as expected
      Edge worker as well as task runner tried via @tenacity.retry() to retry 10 times which took ~9minutes. Then gave-up and terminated.
      If the webserver/connection recovers within the retry attempts the task will continue and all problems are masked.
      In my case I did not restart webserver. So Edge Worker and task runner terminated/gave-up calling.
      In the scheduler the running task was reported as zombie as exected.

As all Executors work on the common DB table and API calls go indirect via DB... in the Executor there is no real ownership needed in my view. Also no adoption. Based on the DB table Executors just report back the "running" tasks for the planning of PARALLELISM.

Would this be OK for you? Or do you see an task adoption is needed?

Reply:
Fixed adoption problem with https://github.com/apache/airflow/pull/42048/commits/6011d7ec7d2cd1506ff23169ea72d9a8313aa3e7

Anything that you want to have changed or OK to approve/merge in the current state?

Reply:
> Will check for the third case. Probably this is where task adoption kicks-in :-D

Yupp :) 

> Fixed adoption problem with https://github.com/apache/airflow/commit/6011d7ec7d2cd1506ff23169ea72d9a8313aa3e7

> Anything that you want to have changed or OK to approve/merge in the current state?

Looking now!

Reply:
Yeah, scheduler stopping shouldn't terminate the tasks -- Celery and Kube executor don't do that for instance.

Reply:
Then... I assume we are okay here.